energymeters/integrationpluginenergymeters.cpp

```diff
@@ -0,0 +1,394 @@
+            if (meter->modbusRtuMasterUuid() == modbusUuid) {
+                Thing *thing = m_energyMeters.key(meter);
+                if (!thing)
+                    return;
+
+                qCWarning(dcEnergyMeters()) << "Modbus RTU hardware resource removed for" << thing << ". The thing will not be functional any more until a new resource has been configured for it.";
+                thing->setStateValue(m_connectionStateTypeIds[thing->thingClassId()], false);
+            }
+        }
+    }); */
+}
+
+void IntegrationPluginEnergyMeters::discoverThings(ThingDiscoveryInfo *info)
+{
+    qCDebug(dcEnergyMeters()) << "Discover things";
+    QList<ThingDescriptor> thingDescriptors;
+
+    if (hardwareManager()->modbusRtuResource()->modbusRtuMasters().isEmpty()) {
+        info->finish(Thing::ThingErrorHardwareNotAvailable, QT_TR_NOOP("No Modbus RTU interface available."));
+        return;
+    }
+
+    if (m_connectionStateTypeIds.contains(info->thingClassId())) {
+        int slaveAddress = info->params().paramValue(m_discoverySlaveAddressParamTypeIds.value(info->thingClassId())).toInt();
+        if (slaveAddress > 254 || slaveAddress == 0) {
+            info->finish(Thing::ThingErrorInvalidParameter, tr("Modbus slave address must be between 1 and 254"));
+            return;
+        }
+        Q_FOREACH(ModbusRtuMaster *modbusMaster, hardwareManager()->modbusRtuResource()->modbusRtuMasters()) {
+            qCDebug(dcEnergyMeters()) << "Found RTU master resource" << modbusMaster << "connected" << modbusMaster->connected();
+            if (!modbusMaster->connected()) {
+                continue;
+            }
+            ThingDescriptor descriptor(info->thingClassId(), QT_TR_NOOP("Energy meter"), QT_TR_NOOP("Slave address ") +QString::number(slaveAddress)+" "+modbusMaster->serialPort());
+            ParamList params;
+            params << Param(m_slaveIdParamTypeIds.value(info->thingClassId()), slaveAddress);
+            params << Param(m_modbusUuidParamTypeIds.value(info->thingClassId()), modbusMaster->modbusUuid());
+            descriptor.setParams(params);
+            info->addThingDescriptor(descriptor);
+        }
+        info->finish(Thing::ThingErrorNoError);
+        return;
+    } else {
+        Q_ASSERT_X(false, "discoverThings", QString("Unhandled thingClassId: %1").arg(info->thingClassId().toString()).toUtf8());
+    }
+}
+
+void IntegrationPluginEnergyMeters::setupThing(ThingSetupInfo *info)
+{
+    Thing *thing = info->thing();
+    qCDebug(dcEnergyMeters()) << "Setup thing" << thing->name();
+
+    if (m_connectionStateTypeIds.contains(thing->thingClassId())) {
+
+        if (m_energyMeters.contains(thing)) {
+            qCDebug(dcEnergyMeters()) << "Setup after rediscovery, cleaning up ...";
+            m_energyMeters.take(thing)->deleteLater();
+        }
+        int address = thing->paramValue(m_slaveIdParamTypeIds.value(thing->thingClassId())).toInt();
```

Comment:
Same as in discovery: if you make the param Uint you don't need to handle negative slave addresses